install/uninstall.ps1

```diff
@@ -0,0 +1,125 @@
+# Enable a default -Verbose flag for debug output.
+[CmdletBinding()]
+
+# Halt script execution at the first failed command.
+$script:ErrorActionPreference='Stop'
+
+$azureauthDefaultLocation = ([System.IO.Path]::Combine($Env:LOCALAPPDATA, "Programs", "AzureAuth"))
+
+if (![string]::IsNullOrEmpty($Env:AZUREAUTH_INSTALL_DIRECTORY) `
+        -And $Env:AZUREAUTH_INSTALL_DIRECTORY -ne $azureauthDefaultLocation) {
+    Write-Warning "Ignoring AZUREAUTH_INSTALL_DIRECTORY environment variable. Custom uninstallation locations are not supported."
+}
+
+function Uninstall {
+    Close-AzureauthInstances
+
+    $locations = Get-AzureauthsInPath
+    Remove-AzureauthDirectories $locations
+    Remove-FromPath $locations
+
+    Write-Output "Uninstalled AzureAuth!"
+}
+
+function Get-AzureauthsInPath {
+    $allLocations = [System.Collections.ArrayList]@()
+    $azureauthsInPath = (Get-Command -Name azureauth -ErrorAction SilentlyContinue -CommandType Application -All).Source
+
+    # Uninstallation is only supported from the default AzureAuth location.
+    # We warn the user of any custom locations that are listed in the PATH.
+    # Installations in custom locations that are not listed in PATH cannot
+    # be found and uninstalled.
+    # We still...
+    ForEach($az in $azureauthsInPath) {
+        $tempDirectory = (Get-Item $az).Directory.FullName
+        if (!$az.Contains($azureauthDefaultLocation)) {
+            Write-Warning "Additional installation found in ${tempDirectory}"
+        } else {
+            # We add the PATH location to the list and not the default location to later
+            # remove it from the PATH (default location is the whole parent folder).
+            $_ = $allLocations.Add($tempDirectory)
+        }
+    }
+
+    return $allLocations
+}
+
+function Remove-AzureauthDirectories {
+    param([System.Collections.ArrayList]$locationsToDelete)
+
+    ForEach ($location in $locationsToDelete) {
+        Remove-InstallationFolder $location
+    }
+    
+    # We also delte AzureAuth parent folder
+    Remove-InstallationFolder $azureauthDefaultLocation
+}
+
+function Remove-InstallationFolder {
+    param ([string]$directory)
+
+    if (Test-Path -Path $directory) {
+        Write-Verbose "Removing installations at '${directory}'"
+        Remove-Item -Force -Recurse $directory
+    } else {
+        Write-Verbose "There were no installations found at '${directory}'"

```

Comment:
I am wondering if we will ever enter this `else` statement. We created a list of azureauth directories that exist already. Am I missing an edge case?

Reply:
User could manually delete the AzureAuth folder and not from the PATH. Though I could change the comment to "There were no installations found at '${directory}', removing reference from PATH"

Reply:
I think @Emilio0404's correct about this edge case. Just because we discover something on the user's `PATH` doesn't mean that it's actually backed by anything on disk still.

I would be wary of crossing purposes in your output though. Saying that you're removing a reference from `PATH` in the `Remove-InstallationFolder` function assumes knowledge from another function or how the whole pipeline is put together. If you're truly making these functions individual components then this function shouldn't be responsible for that message.

Logging is also part of your API, not just the return value.

Reply:
Left the comment as is.